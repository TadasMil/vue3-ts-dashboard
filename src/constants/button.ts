export const ButtonColors = {
  PRIMARY: 'primary',
  SECONDARY: 'secondary',
  TERTIARY: 'tertiary',
  HIGHLIGHT: 'highlight'
} as const

export const ButtonDensities = {
  DEFAULT: 'default',
  COMFORTABLE: 'comfortable',
  COMPACT: 'compact',
  NONE: 'none'
} as const

export const ButtonVariants = {
  DEFAULT: 'default',
  OUTLINED: 'outlined',
  TEXT: 'text'
} as const

export const ButtonRoundeds = {
  SM: 'sm',
  MD: 'md',
  LG: 'lg'
} as const
