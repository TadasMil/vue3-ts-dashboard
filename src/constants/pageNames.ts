export default {
  /* User Related Pages */
  UserDashboard: 'UserDashboard',
  UserRecruitment: 'UserRecruitment',
  UserSchedule: 'UserSchedule',
  UserSupport: 'UserSupport',
  UserSettings: 'UserSettings',
  UserDepartment: 'UserDepartment',
  UserEmployee: 'UserEmployee'
}
