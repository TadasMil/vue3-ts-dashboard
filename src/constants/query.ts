export const QUERY_STATES = {
  IDLE: 'IDLE',
  LOADING: 'LOADING',
  ERROR: 'ERROR',
  NOT_FOUND: 'NOT_FOUND',
  NO_RESULTS: 'NO_RESULTS',
  NOT_AUTHORIZED: 'NOT_AUTHORIZED'
} as const
