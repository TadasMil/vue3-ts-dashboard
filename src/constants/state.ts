export const STATES = {
  LOADING: 'LOADING',
  IDLE: 'IDLE',
  ERROR: 'ERROR',
  NOT_FOUND: 'NOT_FOUND',
  NO_RESULTS: 'NO_RESULTS',
  NOT_AUTHORIZED: 'NOT_AUTHORIZED'
} as const
