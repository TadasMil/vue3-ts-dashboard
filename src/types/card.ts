export interface ICardClasses {
  root: string
  header: string
  body: string
  title: string
  subtitle: string
  content: string
  footer: string
}

export interface ICardPassThroughOptions {
  root?: Record<string, unknown>
  header?: Record<string, unknown>
  body?: Record<string, unknown>
  title?: Record<string, unknown>
  subtitle?: Record<string, unknown>
  content?: Record<string, unknown>
  footer?: Record<string, unknown>
}
