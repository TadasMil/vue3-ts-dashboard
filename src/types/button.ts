import { ButtonColors, ButtonDensities, ButtonRoundeds, ButtonVariants } from '@/constants/button'

export type ButtonColor = (typeof ButtonColors)[keyof typeof ButtonColors]
export type ButtonDensity = (typeof ButtonDensities)[keyof typeof ButtonDensities]
export type ButtonVariant = (typeof ButtonVariants)[keyof typeof ButtonVariants]
export type ButtonRounded = (typeof ButtonRoundeds)[keyof typeof ButtonRoundeds]
