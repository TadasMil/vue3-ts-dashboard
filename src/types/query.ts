import { STATES } from '@/constants/state'

export type IQueryState = (typeof STATES)[keyof typeof STATES]
