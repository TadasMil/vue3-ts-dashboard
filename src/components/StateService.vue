<script setup lang="ts" name="StateService">
import StateCardError from '@/components/StateCardError.vue'
import StateCardLoading from '@/components/StateCardLoading.vue'

import { IQueryState } from '@/types/query'

const { stateIs } = defineProps<{
  stateIs: (active: IQueryState) => boolean
}>()
</script>

<template>
  <slot name="loading" v-if="stateIs('LOADING')">
    <StateCardLoading />
  </slot>

  <slot name="error" v-else-if="stateIs('ERROR')">
    <StateCardError
      title="500"
      subtitle="Internal Server Error"
      message="Sorry, something went wrong."
    />
  </slot>

  <slot name="not_authorized" v-else-if="stateIs('NOT_AUTHORIZED')">
    <StateCardError title="403" subtitle="You are not authorized to access this information." />
  </slot>

  <slot name="not_found" v-else-if="stateIs('NOT_FOUND')">
    <StateCardError title="404" subtitle="The following data could not be found." />
  </slot>

  <slot name="no_results" v-else-if="stateIs('NO_RESULTS')">
    <StateCardError
      subtitle="Sorry, no results found."
      message="Please amend the search criteria or filter field to see all items."
    />
  </slot>

  <slot v-else-if="stateIs('IDLE')" />

  <p v-else class="bg-red-5 text-9xl text-white">State is not a valid state.</p>
</template>
