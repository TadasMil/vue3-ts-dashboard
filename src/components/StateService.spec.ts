import { render } from '@testing-library/vue'
import { describe, expect, it } from 'vitest'

import { STATES } from '@/constants/query'

import { IQueryState } from '@/types/query'

import '@testing-library/jest-dom'
import StateService from './StateService.vue'

describe('StateService.vue', () => {
  const testCases = [
    { state: STATES.LOADING, slot: 'loading' },
    { state: STATES.ERROR, slot: 'error' },
    { state: STATES.NOT_AUTHORIZED, slot: 'not_authorized' },
    { state: STATES.NOT_FOUND, slot: 'not_found' },
    { state: STATES.NO_RESULTS, slot: 'no_results' }
  ]

  testCases.forEach(({ state }) => {
    it(`renders ${state} component`, async () => {
      const { html } = render(StateService, {
        props: { stateIs: (s: IQueryState) => s === state }
      })

      expect(html()).toMatchSnapshot()
    })
  })

  testCases.forEach(({ state, slot }) => {
    it(`renders ${slot} slot when state is ${state}`, async () => {
      const { queryByTestId } = render(StateService, {
        props: { stateIs: (s: IQueryState) => s === state },
        slots: { [slot]: `<div data-testid="${slot}-slot"></div>` }
      })

      expect(queryByTestId(`${slot}-slot`)).toBeInTheDocument()
    })
  })

  it('renders error message for an invalid state', async () => {
    const { getByText } = render(StateService, {
      props: { stateIs: (state: IQueryState) => state === ('INVALID_STATE' as any) }
    })

    expect(getByText('State is not a valid state.')).toBeInTheDocument()
  })
})
