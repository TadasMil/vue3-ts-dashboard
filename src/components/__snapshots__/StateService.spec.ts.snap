// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`StateService.vue > renders ERROR component 1`] = `
"<div data-v-99b8d817=\\"\\" data-v-b435c541=\\"\\" class=\\"state-container\\">
  <div data-v-b435c541=\\"\\" class=\\"error-card\\">
    <p data-v-b435c541=\\"\\" class=\\"error-card__title\\">500</p>
    <p data-v-b435c541=\\"\\" class=\\"error-card__subtitle\\">Internal Server Error</p>
    <p data-v-b435c541=\\"\\" class=\\"error-card__message\\">Sorry, something went wrong.</p>
  </div>
</div>"
`;

exports[`StateService.vue > renders LOADING component 1`] = `
"<div data-v-99b8d817=\\"\\" data-v-cbf0e5c2=\\"\\" class=\\"state-container\\"><svg data-v-cbf0e5c2=\\"\\" viewBox=\\"0 0 24 24\\" width=\\"1.2em\\" height=\\"1.2em\\" class=\\"loading-spinner\\">
    <path fill=\\"currentColor\\" d=\\"M12 4V2A10 10 0 0 0 2 12h2a8 8 0 0 1 8-8Z\\"></path>
  </svg></div>"
`;

exports[`StateService.vue > renders NO_RESULTS component 1`] = `
"<div data-v-99b8d817=\\"\\" data-v-b435c541=\\"\\" class=\\"state-container\\">
  <div data-v-b435c541=\\"\\" class=\\"error-card\\">
    <!--v-if-->
    <p data-v-b435c541=\\"\\" class=\\"error-card__subtitle\\">Sorry, no results found.</p>
    <p data-v-b435c541=\\"\\" class=\\"error-card__message\\">Please amend the search criteria or filter field to see all items.</p>
  </div>
</div>"
`;

exports[`StateService.vue > renders NOT_AUTHORIZED component 1`] = `
"<div data-v-99b8d817=\\"\\" data-v-b435c541=\\"\\" class=\\"state-container\\">
  <div data-v-b435c541=\\"\\" class=\\"error-card\\">
    <p data-v-b435c541=\\"\\" class=\\"error-card__title\\">403</p>
    <p data-v-b435c541=\\"\\" class=\\"error-card__subtitle\\">You are not authorized to access this information.</p>
    <!--v-if-->
  </div>
</div>"
`;

exports[`StateService.vue > renders NOT_FOUND component 1`] = `
"<div data-v-99b8d817=\\"\\" data-v-b435c541=\\"\\" class=\\"state-container\\">
  <div data-v-b435c541=\\"\\" class=\\"error-card\\">
    <p data-v-b435c541=\\"\\" class=\\"error-card__title\\">404</p>
    <p data-v-b435c541=\\"\\" class=\\"error-card__subtitle\\">The following data could not be found.</p>
    <!--v-if-->
  </div>
</div>"
`;
